lib/ansible/modules/storage/netapp/na_ontap_qtree.py

```diff
@@ -0,0 +1,233 @@
+            'qtree-info', **{'vserver': self.vserver,
+                             'volume': self.flexvol_name,
+                             'qtree': self.name})
+
+        query = netapp_utils.zapi.NaElement('query')
+        query.add_child_elem(query_details)
+        qtree_list_iter.add_child_elem(query)
+
+        result = self.server.invoke_successfully(qtree_list_iter,
+                                                 enable_tunneling=True)
+
+        if (result.get_child_by_name('num-records') and
+                int(result.get_child_content('num-records')) >= 1):
+            return True
+        else:
+            return False
+
+    def create_qtree(self):
+        qtree_create = netapp_utils.zapi.NaElement.create_node_with_children(
+            'qtree-create', **{'volume': self.flexvol_name,
+                               'qtree': self.name})
+
+        try:
+            self.server.invoke_successfully(qtree_create,
+                                            enable_tunneling=True)
+        except netapp_utils.zapi.NaApiError as e:
+            self.module.fail_json(msg="Error provisioning qtree %s: %s" % (self.name, to_native(e)),
+                                  exception=traceback.format_exc())
+
+    def delete_qtree(self):
+        path = '/vol/%s/%s' % (self.flexvol_name, self.name)
+        qtree_delete = netapp_utils.zapi.NaElement.create_node_with_children(
+            'qtree-delete', **{'qtree': path})
+
+        try:
+            self.server.invoke_successfully(qtree_delete,
+                                            enable_tunneling=True)
+        except netapp_utils.zapi.NaApiError as e:
+            self.module.fail_json(msg="Error deleting qtree %s: %s" % (path, to_native(e)),
+                                  exception=traceback.format_exc())
+
+    def rename_qtree(self):
+        path = '/vol/%s/%s' % (self.flexvol_name, self.name)
+        new_path = '/vol/%s/%s' % (self.flexvol_name, self.new_name)
+        qtree_rename = netapp_utils.zapi.NaElement.create_node_with_children(
+            'qtree-rename', **{'qtree': path,
+                               'new-qtree-name': new_path})
+
+        try:
+            self.server.invoke_successfully(qtree_rename,
+                                            enable_tunneling=True)
+        except netapp_utils.zapi.NaApiError as e:
+            self.module.fail_json(msg="Error renaming qtree %s: %s" % (self.name, to_native(e)),
+                                  exception=traceback.format_exc())
+
+    def apply(self):
+        changed = False
+        qtree_exists = False
+        rename_qtree = False
+        netapp_utils.ems_log_event("na_ontap_qtree", self.server)
+        qtree_detail = self.get_qtree()
+        if qtree_detail:
+            qtree_exists = True
+            if self.state == 'absent':  # delete
+                changed = True
+            else:  # rename
+                if self.new_name and self.name != self.new_name:
```

Comment:
not idempotent; if `new_name` specified, check for existence of qtree again with `new_name` before determining it needs to be created; if exists with new_name, return `changed: false`

Reply:
We're expecting the rename command to fail if the qtree with new_name already exists.